dynamax/linear_gaussian_ssm/parallel_inference.py

```diff
@@ -21,75 +21,41 @@
 
 Dynamax
 
-      F₀,Q₀           F₁,Q₁         F₂,Q₂
+      F₁,Q₁           F₂,Q₂         F₃,Q₃
 Z₀ ─────────── Z₁ ─────────── Z₂ ─────────── Z₃ ─────...
 |              |              |              |
 | H₀,R₀        | H₁,R₁        | H₂,R₂        | H₃,R₃
 |              |              |              |
 Y₀             Y₁             Y₂             Y₃ 
 
-'''
+"""
 
 import jax.numpy as jnp
 from jax import vmap, lax
 from jaxtyping import Array, Float
-from typing import NamedTuple
+from typing import NamedTuple, Optional
 from dynamax.types import PRNGKey
 from functools import partial
 import warnings
 
 from tensorflow_probability.substrates.jax.distributions import (
     MultivariateNormalDiagPlusLowRankCovariance as MVNLowRank,
-    MultivariateNormalFullCovariance as MVN)
+    MultivariateNormalFullCovariance as MVN,
+)
 
 from jax.scipy.linalg import cho_solve, cho_factor
 from dynamax.utils.utils import symmetrize, psd_solve
 from dynamax.linear_gaussian_ssm import PosteriorGSSMFiltered, PosteriorGSSMSmoothed, ParamsLGSSM
+from dynamax.linear_gaussian_ssm.inference import preprocess_args, _get_one_param, _get_params, _log_likelihood
 
 
-def _get_one_param(x, dim, t):
-    """Helper function to get one parameter at time t."""
-    if callable(x):
-        return x(t)
-    elif x.ndim == dim + 1:
-        return x[t]
-    else:
-        return x
-
-def _get_params(params, num_timesteps, t):
-    """Helper function to get parameters at time t."""
-    assert not callable(params.emissions.cov), "Emission covariance cannot be a callable."
-
-    F = _get_one_param(params.dynamics.weights, 2, t)
-    b = _get_one_param(params.dynamics.bias, 1, t)
-    Q = _get_one_param(params.dynamics.cov, 2, t)
-    H = _get_one_param(params.emissions.weights, 2, t+1)
-    d = _get_one_param(params.emissions.bias, 1, t+1)
-
-    if len(params.emissions.cov.shape) == 1: 
-        R = _get_one_param(params.emissions.cov, 1, t+1)
-    elif len(params.emissions.cov.shape) > 2: 
-        R = _get_one_param(params.emissions.cov, 2, t+1)
-    elif params.emissions.cov.shape[0] != num_timesteps:
-        R = _get_one_param(params.emissions.cov, 2, t+1)
-    elif params.emissions.cov.shape[1] != num_timesteps:
-        R = _get_one_param(params.emissions.cov, 1, t+1)
-    else:
-        R = _get_one_param(params.emissions.cov, 2, t+1)
-        warnings.warn(
-            "Emission covariance has shape (N,N) where N is the number of timesteps. "
-            "The covariance will be interpreted as static and non-diagonal. To "
-            "specify a dynamic and diagonal covariance, pass it as a 3D array.")
-
-    return F, b, Q, H, d, R
```

Comment:
I remoted these kind of duplicated utility functions in `parallel_inference.py` and used the one defined in `inference.py`